src/Surfnet/StepupGateway/SecondFactorOnlyBundle/Controller/SecondFactorOnlyController.php

```diff
@@ -0,0 +1,261 @@
+        $loaId = $this->verifyAuthnContextClassRef($authnContextClassRef, $logger);
+        if (!$loaId) {
+            /** @var \Surfnet\StepupGateway\SecondFactorOnlyBundle\Service\ResponseRenderingService $responseRendering */
+            $responseRendering = $this->get('second_factor_only.response_rendering');
+            return $responseRendering->renderRequesterFailureResponse(
+                $this->getResponseContext()
+            );
+        }
+        $stateHandler->setRequiredLoaIdentifier($loaId);
+
+        $logger->notice(
+            'Forwarding to second factor controller for loa determination and handling'
+        );
+        return $this->forward(
+            'SurfnetStepupGatewayGatewayBundle:SecondFactor:selectSecondFactorForVerification'
+        );
+    }
+
+    /**
+     * @param string $authnContextClassRef
+     * @param LoggerInterface $logger
+     * @return string
+     */
+    private function verifyAuthnContextClassRef(
+        $authnContextClassRef,
+        LoggerInterface $logger
+    ) {
+        if (!$authnContextClassRef) {
+            $logger->info(
+                'No LOA requested, sending response with status Requester Error'
+            );
+            return;
+        }
+
+        /** @var LoaAliasLookupService $loaAliasLookup */
+        $loaAliasLookup = $this->get('second_factor_only.loa_alias_lookup');
+        $loaId = $loaAliasLookup->findLoaIdByAlias($authnContextClassRef);
+
+        if (!$loaId) {
+            $logger->info(sprintf(
+                'Requested required Loa "%s" does not have a second factor alias,'
+                .' sending response with status Requester Error',
+                $authnContextClassRef
+            ));
+            return;
+        }
+
+        $loaResolutionService = $this->get('surfnet_stepup.service.loa_resolution');
+
+        if (!$loaResolutionService->hasLoa($loaId)) {
+            $logger->info(sprintf(
+                'Requested required Loa "%s" does not exist,'
+                .' sending response with status Requester Error',
+                $authnContextClassRef
+            ));
+            return;
+        }
+
+        return $loaId;
+    }
+
+    /**
+     * @param string $spEntityId
+     * @param string $nameId
+     * @param LoggerInterface $logger
+     * @return bool
+     */
+    private function verifyNameId($spEntityId, $nameId, LoggerInterface $logger)
+    {
+        if (!$nameId) {
+            $logger->info(
+                'No NameID provided, sending response with status Requester Error'
+            );
+            return false;
+        }
+
+        $entityService = $this->get('second_factor_only.entity_service');
+        $serviceProvider = $entityService->getServiceProvider($spEntityId);
+
+        if (!$serviceProvider->isAllowedToUseSecondFactorOnlyFor($nameId)) {
+            $logger->info(
+                'No NameID provided, sending response with status Requester Error'
```

Comment:
Some copy-pasta here :wink: I think there's also merit to logging that the NameID _is_ allowed to use second-factor-only.